src/Microsoft.AspNet.WebUtilities/Encoders/HtmlEncoder.cs

```diff
@@ -0,0 +1,229 @@
+    /// can be represented unencoded.
+    /// </summary>
+    /// <remarks>
+    /// Once constructed, instances of this class are thread-safe for multiple callers.
+    /// </remarks>
+    public unsafe sealed class HtmlEncoder
+    {
+        // The default HtmlEncoder (Basic Latin), instantiated on demand
+        private static HtmlEncoder _defaultEncoder;
+
+        // A bitmap of characters which are allowed to be returned unescaped.
+        private readonly uint[] _allowedCharsBitmap = new uint[0x10000 / 32];
+
+        /// <summary>
+        /// Instantiates an encoder using the 'Basic Latin' code table as the allow list.
+        /// </summary>
+        public HtmlEncoder()
+            : this(UnicodeFilters.BasicLatin)
+        {
+        }
+
+        /// <summary>
+        /// Instantiates an encoder using a custom allow list of characters.
+        /// </summary>
+        public HtmlEncoder(params ICodePointFilter[] filters)
+        {
+            if (filters == null)
+            {
+                return; // no characters are allowed, just no-op immediately
+            }
+
+            // Punch a hole for each allowed code point across all filters (this is an OR).
+            // We don't allow supplementary (astral) characters for now.
+            foreach (var filter in filters)
+            {
+                foreach (var codePoint in filter.GetAllowedCodePoints())
+                {
+                    if (!UnicodeHelpers.IsSupplementaryCodePoint(codePoint))
+                    {
+                        AllowCharacter((char)codePoint);
+                    }
+                }
+            }
+
+            // Forbid characters that are special in HTML
+            ForbidCharacter('<');
+            ForbidCharacter('>');
+            ForbidCharacter('&');
+            ForbidCharacter('\''); // can be used to escape attributes
+            ForbidCharacter('\"'); // can be used to escape attributes
+            ForbidCharacter('+'); // technically not HTML-specific, but can be used to perform UTF7-based attacks
+
+            // Forbid codepoints which aren't mapped to characters or which are otherwise always disallowed
+            // (includes categories Cc, Cs, Co, Cn, Zl, Zp)
+            uint[] definedCharactersBitmap = UnicodeHelpers.GetDefinedCharacterBitmap();
+            Debug.Assert(definedCharactersBitmap.Length == _allowedCharsBitmap.Length);
+            for (int i = 0; i < _allowedCharsBitmap.Length; i++)
+            {
+                _allowedCharsBitmap[i] &= definedCharactersBitmap[i];
+            }
+        }
+
+        /// <summary>
+        /// A default instance of the HtmlEncoder, equivalent to allowing only
+        /// the 'Basic Latin' character range.
+        /// </summary>
+        public static HtmlEncoder Default
+        {
+            get
+            {
+                HtmlEncoder defaultEncoder = Volatile.Read(ref _defaultEncoder);
```

Comment:
Very much a case-by-case thing that depends on how many occurrences of LazyInitializer there are and how often the accessor is likely to be called. For instance, I just stuck with a standard LazyInitializer on UnicodeFilters.None for simplicity.